csharp/src/Google.Protobuf/Collections/MapField.cs

```diff
@@ -221,17 +227,15 @@ public bool TryGetValue(TKey key, out TValue value)
             }
         }
 
-        // TODO: Make these views?
-
         /// <summary>
         /// Gets a collection containing the keys in the map.
         /// </summary>
-        public ICollection<TKey> Keys { get { return list.Select(t => t.Key).ToList(); } }
+        public ICollection<TKey> Keys { get { return new MapView<TKey>(this, pair => pair.Key, ContainsKey); } }
```

Comment:
I think BCL collections actually cache the view once they lazy-initialize it.

See here:
http://referencesource.microsoft.com/#mscorlib/system/collections/hashtable.cs,836

Also below for Values.